### PR TITLE
feat: add support for drawio.png

### PR DESCRIPTION
**Pull request by Saphnx**
Add support for editable png diagrams.

### CODE REVIEW

File: mkdocs_drawio/plugin.py [hunk 1]
```diff
@@ -195,7 +197,12 @@ def substitute_with_url(config: Dict, url: str) -> str:
     @staticmethod
     def substitute_with_file(config: Dict, path: Path, src: str, page: str) -> str:
         try:
-            diagram_xml = etree.parse(path.joinpath(src).resolve())
+            if src.endswith(".png"):
+                img = Image.open(path.joinpath(path, src))
+                xml_data = unquote(img.info.get("mxfile"))
+                diagram_xml = etree.fromstring(xml_data.encode())
```
[the remark this answers is not in the document]
Comment (Saphnx) [reply mkdocs_drawio/plugin.py]: If it's missing any mxfile data, it'll fail to retrieve the metadata and therefore won't be a valid diagram, causing it to fall back to the existing error handling. I've updated the example to include this case:

image

I've also updated mkdocs-classic, as I'd missed that and noticed it when I was pulling in the latest changes.

[the remark this answers is not in the document]
Comment (tuunit) [reply mkdocs_drawio/plugin.py]: I think this is undesirable as it could easily happen that users have named files diagram.drawio.png without it containing the mxdata in the png metadata and this PR would hence be a breaking change for those users of this plugin. Can we introduce a special error handling for png files to force rendering/treating them as normal images if the parsing fails?

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -1,3 +1,4 @@
 FROM mcr.microsoft.com/vscode/devcontainers/python:3-bullseye
 
 RUN pip3 install poetry wheel mkdocs mkdocs-material mkdocs-print-site-plugin BeautifulSoup4
+RUN pip3 install poetry wheel mkdocs mkdocs-material mkdocs-print-site-plugin BeautifulSoup4 Pillow
```
Comment (tuunit) [inline .devcontainer/Dockerfile]: why does this duplicate the previous line instead of just adding Pillow?

File: mkdocs_drawio/plugin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Saphnx) [reply mkdocs_drawio/plugin.py]: Good point, I hadn't considered that people may do that without intentionally saving it as an editable PNG. In this case, it's a valid flow, so I've shifted things around a little and introduced a warning for when the mxfile data is missing on a .drawio.png file.

The flow now is:

1. Attempt to extract mxfile xml from file using new _retrieve_mxfile_ method:
    1. If PNG with no mxfile data, log a warning and return None
    2. Else, parse the mxfile data into XML and return it
    3. An exception will be thrown if it fails to read or parse the file
2. Handle result:
    1. If None, it's a PNG so simply skip this diagram to allow it to render normally
    2. If an exception was thrown, log an error and replace the diagram with "Not a diagram file" message (as before)
    3. If it successfully retrieved the XML, continue
3. Call _substitute_with_file_ method as before, but this time passing in the XML instead of the path and src:
    1. If successful, return the substitution
    2. If an exception was thrown, as in step 2, log an error and replace the diagram with "Not a diagram file" message (as before)

Does the flow make sense? Or did you have something else in mind? I have little experience with Python, so happy to tweak things if it can be tidied up further.

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (Saphnx) [reply .devcontainer/Dockerfile]: Ah, apologies! That's from the merge after the two PR's, I missed that. Thanks!

File: pyproject.toml [hunk 3]
```diff
@@ -26,6 +26,7 @@ include = [
 python = ">=3.8.0,<4.0"
 requests = ">=2.0"
 Jinja2 = ">=3.0"
+Pillow = "<=12.1.1"
```
Comment (tuunit) [inline pyproject.toml]: Pillow is quite excessive for just retrieving a bit of metadata from a PNG. It introduces a significant amount of sub dependencies and C code compilation which easily fails or is just unnecessary overhead in terms of compute and storage. Therefore I suggest we drop pillow in favour of `pypng`

Comment (tuunit) [reply pyproject.toml]: As per PNG Spec RFC2083

https://datatracker.ietf.org/doc/html/rfc2083#page-24

This should work for extracting the mxfile contents without the overhead of pillow 

```python
import png

r = png.Reader(filename="file.png")
for chunk in r.chunks():
    if chunk[0] == b"tEXt":
        key, value = chunk[1].split(b"\x00", 1)
        if key == b"mxfile":
            mxfile = value.decode("latin-1")
```

File: mkdocs_drawio/plugin.py [hunk 4]
```diff
@@ -182,36 +184,71 @@ def render_drawio_diagrams(self, output_content, page):
...
+            # If none exists, handle it gracefully
+            if mxfile_metadata is None:
+                LOGGER.warning(
+                    f"Warning: PNG file '{src}' on path '{path}' missing mxfile metadata"
+                )
+                return None
+
+            xml_data = unquote(mxfile_metadata)
+            return etree.fromstring(xml_data.encode())
+        else:
```
Comment (tuunit) [inline mkdocs_drawio/plugin.py]: this else is redundant just put the return here

Comment (tuunit) [inline mkdocs_drawio/plugin.py]: this else is redundant just put the return here to reduce nesting

Comment (tuunit) [reply mkdocs_drawio/plugin.py]: thinking about it move the return to the verify top and invert the if statement

```python
if not src.endswith(".png"):
  etree.parse(path.joinpath(src).resolve())
```

this way we adhere to early return principles and you can get rid of the indentation for the complete png extraction logic

Comment (tuunit) [reply mkdocs_drawio/plugin.py]: furthermore the regex ignores the casing this statement is case sensitive so convert to lower casing first and then check the contains

**Review by tuunit — CHANGES_REQUESTED** [review]

File: mkdocs_drawio/plugin.py (hunk 4 above)
Comment (Saphnx) [reply mkdocs_drawio/plugin.py]: Well that's quite a blatant case, gave me a good laugh at myself at least! Good spot, and apologies, I must have overlooked it in my moving things around and working out the best flow.

Comment (Saphnx) [reply mkdocs_drawio/plugin.py]: > thinking about it move the return to the verify top and invert the if statement
> 
> ```python
> if not src.endswith(".png"):
>   etree.parse(path.joinpath(src).resolve())
> ```
> 
> this way we adhere to early return principles and you can get rid of the indentation for the complete png extraction logic

Absolutely - much nicer.

File: pyproject.toml (hunk 3 above)
Comment (Saphnx) [reply pyproject.toml]: This is much better, thanks for your effort =)

File: mkdocs_drawio/plugin.py (hunk 4 above)
Comment (Saphnx) [reply mkdocs_drawio/plugin.py]: > furthermore the regex ignores the casing this statement is case sensitive so convert to lower casing first and then check the contains

Ahh good spot, thanks for that. That'd be the sort of thing I'd cover with a unit test normally, but there's none in the project so far. Is that a preference given the size of the plugin? Or would you be happy if I added some? Not done it with Python before, but I see _Unittest_ is built-in?

Comment (Saphnx) [reply mkdocs_drawio/plugin.py]: Hey @tuunit, any feeling on the above? Or are you happy to merge as-is?

Comment (tuunit) [reply mkdocs_drawio/plugin.py]: Hey @Saphnx sorry completely lost track of this PR. I was on a long vacation and switched jobs, had a lot going on 😅 

PR LGTM now

**Review by tuunit — APPROVED** [review]

**Comment by Saphnx** [comment]
No worries on the holiday! We all have other things going on =)

Unfortunately, it looks like something is going awry when _mkdocs build_ is run (at least with what I'm working with now). I'll have a look at the issue when I get home.

In the meantime, I've raised an issue so anyone encountering it knows to pin to version 1.15.0 to get it working again.
